### PR TITLE
Fusion level axioms

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes https://github.com/NVIDIA/Fuser/issues/320. Add a new optimization pass that append `extent > 0` as axioms of the fusion.

Currently, we are not doing concretization of extent zero, so this assumption is actually wrong. But fortunately, it is not too wrong, given the fact that:
- no test / benchmark fail
- AFAICT, except at parallel dimension map (where we already assume `extent > 0`), the only difference that `extent > 0` vs `extent >= 0` is when `extent` is a denominator. Considering that we already have `preserve_error = false` by default, so changing from `extent >= 0` to `extent > 0` will hopefully be a no-op for lowering.

But this do help on vectorization analysis. Currently, I am seeing a lot of things like `where(abs(i1) == i1, i2, i3)`, and the added axioms should help on eliminate this trivial computation.

Also note that I removed the explicit assumption on parallel dimension map, which means that, if we are using fusion executor manually, we must run pre-segmentation pass before scheduling in order to get the same behavior as using executor cache. In the future, I think this is the direction to go as we are adding more and more pre-segmentation passes. Especially, I am thinking about moving `replaceSymbolicSizes` from a lowering pass into a pre-segmentation or maybe pre-concretization pass. This would reduce the number of `Val*` and generally speeding up all expr simplification and `sameAs` check.

I am OK with holding this PR until zero extent concretization is merged. But I don't think this is necessary. (Just be a bad guy and break things😈)

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Review by naoyam — APPROVED** [review]
Agree this should be relatively safe. I still think we should consider a different IterType for size-zero domains.

**Comment by zasdfgbnm** [comment]
@naoyam Clang-tidy is broken due to unrelated thing:
https://github.com/NVIDIA/Fuser/blob/1ca858158ffbfab3c6c756e90d0c1c4a7137f946/.github/workflows/lint.yml#L60-L63
I think I need admin power to force merge.

**Comment by zasdfgbnm** [comment]
And I agree with a separate IterType for zero extent.

**Comment by zasdfgbnm** [comment]
> @naoyam Clang-tidy is broken due to unrelated thing:
> 
> https://github.com/NVIDIA/Fuser/blob/1ca858158ffbfab3c6c756e90d0c1c4a7137f946/.github/workflows/lint.yml#L60-L63
> 
> 
> I think I need admin power to force merge.

Maybe not, I think the clang-tidy error is easy to fix. Let me try it.

File: .github/workflows/lint.yml
Comment (zasdfgbnm) [inline .github/workflows/lint.yml:1]: cc @rdspring1 @jjsjann123

**Comment by jacobhinkle** [comment]
I think we can guarantee positivity somewhat soon, and kill a few birds with one stone. #449 in a start, and #448 is the followup (haven't started that just yet but hopefully will soon).

**Pull request merged**